include/pybind11/pybind11.h

```diff
@@ -1064,24 +1064,68 @@ template <typename Type, typename... Extra> iterator make_iterator(Type &value,
     return make_iterator(std::begin(value), std::end(value), extra...);
 }
 
+NAMESPACE_BEGIN(detail)
+// These have to be separated like this because the lambda in the proper version won't compile if
+// the C++ types aren't implicitly convertible.
+template <typename InputType, typename OutputType>
+typename std::enable_if<std::is_convertible<InputType, OutputType>::value>::type
+implicitly_cpp_convertible(detail::type_info &input_type) {
+    // Some validity checks on output type:
+    if (!std::is_destructible<OutputType>::value)
+        pybind11_fail("implicitly_convertible: " + type_id<OutputType>() + " is not destructible");
+    if (!std::is_move_constructible<OutputType>::value && !std::is_copy_constructible<OutputType>::value)
+        pybind11_fail("implicitly_convertible: " + type_id<OutputType>() + " is not move- or copy-constructible");
+
+    input_type.implicit_cpp_conversions.emplace(std::type_index(typeid(OutputType)),
+            [](void *input) -> void * {
+                // Takes a pointer to the InputType object, allocates an OutputType with the implicit
+                // conversion value and moves/copies the converted value into it, returning a pointer.
+                // Returns nullptr if input is nullptr.  The caller is responsible for destruction.
+                if (!input) return nullptr;
+                OutputType t = *reinterpret_cast<InputType*>(input);
+                return new OutputType(std::move(t));
+            });
+}
+template <typename InputType, typename OutputType>
+typename std::enable_if<!std::is_convertible<InputType, OutputType>::value>::type
+implicitly_cpp_convertible(detail::type_info &) {
+    pybind11_fail("implicitly_convertible: C++ implicit conversion from " + type_id<InputType>() +
+            " to " + type_id<OutputType>() + " is not valid");
+}
+NAMESPACE_END(detail)
+
 template <typename InputType, typename OutputType> void implicitly_convertible() {
-    auto implicit_caster = [](PyObject *obj, PyTypeObject *type) -> PyObject * {
-        if (!detail::type_caster<InputType>().load(obj, false))
-            return nullptr;
-        tuple args(1);
-        args[0] = obj;
-        PyObject *result = PyObject_Call((PyObject *) type, args.ptr(), nullptr);
-        if (result == nullptr)
-            PyErr_Clear();
-        return result;
-    };
-    auto &registered_types = detail::get_internals().registered_types_cpp;
-    auto it = registered_types.find(std::type_index(typeid(OutputType)));
-    if (it == registered_types.end())
-        pybind11_fail("implicitly_convertible: Unable to find type " + type_id<OutputType>());
-    ((detail::type_info *) it->second)->implicit_conversions.push_back(implicit_caster);
+    auto *output_type = detail::get_type_info(typeid(OutputType));
+    if (output_type) {
+        // Implicit conversion to a registered type, via registered constructors
+        if (!output_type)
```

Comment:
Fixed in f2ad0c8 (removes the if) and afc9800 (throws an error about needing implicit.h when applicable).